### ISSUE TITLE
F-Droid listing

### ISSUE THREAD

**Issue by IzzySoft** [Wavesonics/hammer-editor#191]
First congrats to your listing at F-Droid! At this point, apps are usually removed from my repo (especially when they did not establish reproducible builds) – but your Readme still links to my repo instead of F-Droid's. Will you replace those links?

**Comment by Wavesonics** [comment]
Done, thanks @IzzySoft !

**Comment by IzzySoft** [comment]
Thank you! Will be gone from my repo with the next sync then. All the best!

**Comment by IzzySoft** [comment]
@Wavesonics as our policy concerning removal of apps which have reached F-Droid has changed shortly after our last meet, we've been looking to bring them back. On that path, we were looking if we can bring them back with [Reproducible Builds](https://reproducible-builds.org/) (for some background, see e.g. [Reproducible Builds, special client support and more at IzzyOnDroid](https://android.izzysoft.de/articles/named/iod-rbs-mirrors-clients)). Trying Hammer, we found it's coming pretty close – with just two minor deviations:

```diff
  -rw-r--r--  0.0 unx      120 b-      118 defN 1981-01-01 01:01:02 df7ede7f META-INF/version-control-info.textproto
- -rw-r--r--  0.0 unx     6093 b-     6093 stor 1981-01-01 01:01:02 6bc51433 assets/dexopt/baseline.prof
+ -rw-r--r--  0.0 unx     6093 b-     6093 stor 1981-01-01 01:01:02 efabb843 assets/dexopt/baseline.prof
  -rw-r--r--  0.0 unx      672 b-      672 stor 1981-01-01 01:01:02 84f36814 assets/dexopt/baseline.profm
- -rw-r--r--  0.0 unx  8345692 b-  3704109 defN 1981-01-01 01:01:02 24f0a2b9 classes.dex
+ -rw-r--r--  0.0 unx  8345692 b-  3704108 defN 1981-01-01 01:01:02 26f0e747 classes.dex
  -rw-r--r--  0.0 unx    10096 b-    10096 stor 1981-01-01 01:01:02 9734baa0 lib/arm64-v8a/libandroidx.graphics.path.so
  -rw-r--r--  0.0 unx     7112 b-     7112 stor 1981-01-01 01:01:02 4f56ac38 lib/arm64-v8a/libdatastore_shared_counter.so
  -rw-r--r--  0.0 unx   387256 b-   387256 stor 1981-01-01 01:01:02 cfb9a07b lib/arm64-v8a/librenderscript-toolkit.so
@@ -637,7 +637,7 @@
  -rw----     0.0 fat      792 b-      412 defN 1981-01-01 01:01:02 dfcc1f44 res/M2.xml
  -rw----     0.0 fat      612 b-      301 defN 1981-01-01 01:01:02 3adb1c93 res/M21.xml
  -rw----     0.0 fat     1224 b-      347 defN 1981-01-01 01:01:02 9361e281 res/M4.xml
- -rw----     0.0 fat   122385 b-    14504 defN 1981-01-01 01:01:02 52667113 res/M7.json
+ -rw----     0.0 fat   122386 b-    14506 defN 1981-01-01 01:01:02 53193cda res/M7.json
  -rw----     0.0 fat     1196 b-      397 defN 1981-01-01 01:01:02 91c0d808 res/M7.xml
```

Going bottom-up: `M7.json` is generated by [AboutLibraries](https://github.com/mikepenz/AboutLibraries), and by default also inserts a build timestamp – which is exactly the case here. And as that timestamp is generated at build time, yours and ours differ. Which can be easily avoided by… just telling AboutLibraries to not include it 😜 As you use Kotlin and a `build.gradle.kts`, this would be your snippet for that:

```kotlin
aboutLibraries {
    // Remove the "generated" timestamp to allow for reproducible builds
    excludeFields = arrayOf("generated")
}
```

Next, jumping to the top: `baseline.prof` most likely only differs because `classes.dex` does, as it holds the checksum of the latter. So let's see what difference `classes.dex` has. Quite a lot, to be honest (the diff file is about 150 kB) – but they all look quite similar:

```diff
 |: new-instance v0, Ldev/icerock/moko/resources/ImageResource;
-|: const v1, #float 1.7945e+38 // #7f0700c4
+|: const v1, #float 1.7945e+38 // #7f0700bc
 |: invoke-direct {v0, v1}, Ldev/icerock/moko/resources/ImageResource;.<init>:(I)V
-|: sput-object v0, Lcom/darkrockstudios/apps/hammer/MR$images;.icon_undo:Ldev/icerock/moko/resources/ImageResource;
+|: sput-object v0, Lcom/darkrockstudios/apps/hammer/MR$images;.icon_bold:Ldev/icerock/moko/resources/ImageResource;
 |: new-instance v0, Ldev/icerock/moko/resources/ImageResource;
```

Seems to be all about image resources. Does that ring a bell with you? I know about differences in PNGs generated (or modified) at runtime, such as PNGCrunch or generation from vector-drawables – but such "image resources" are new to me. There's also a huge block which is contained in your build but not in ours, starting with this:

```diff
 |: new-instance v0, Ldev/icerock/moko/resources/StringResource;
-|: const v1, #float 1.91418e+38 // #7f1001da
-|: invoke-direct {v0, v1}, Ldev/icerock/moko/resources/StringResource;.<init>:(I)V
-|: sput-object v0, Lcom/darkrockstudios/apps/hammer/MR$images;.settings_heading:Ldev/icerock/moko/resources/StringResource;
-|: new-instance v0, Ldev/icerock/moko/resources/StringResource;
-|: const v1, #float 1.91419e+38 // #7f100200
-|: invoke-direct {v0, v1}, Ldev/icerock/moko/resources/StringResource;.<init>:(I)V
-|: sput-object v0, Lcom/darkrockstudios/apps/hammer/MR$images;.settings_theme_label:Ldev/icerock/moko/resources/StringResource;
-|: new-instance v0, Ldev/icerock/moko/resources/StringResource;
-|: const v1, #float 1.91418e+38 // #7f1001d8
-|: invoke-direct {v0, v1}, Ldev/icerock/moko/resources/StringResource;.<init>:(I)V
-|: sput-object v0, Lcom/darkrockstudios/apps/hammer/MR$images;.settings_example_project_header:Ldev/icerock/moko/resources/StringResource;
-|: new-instance v0, Ldev/icerock/moko/resources/StringResource;
-|: const v1, #float 1.91419e+38 // #7f1001e4
```

(but then in a completely different place, a similar but not matching part in ours). Could it be the APK was not built from a clean tree (e.g. local modifications, or artifacts remaining from previous builds)?

Would you be interested in having your app restored at IzzyOnDroid as reproducible build, and dig into this with us?

**Comment by Wavesonics** [comment]
@IzzySoft ya I could look into this, I just put in the about libraries fix.

The moko resources stuff will take a little more research

**Comment by IzzySoft** [comment]
Thanks! Let me know when you want another APK tested.

**Comment by IzzySoft** [comment]
Report from v1.10.0 – the diff is even bigger now, and not only moko: [diff.zip](https://github.com/user-attachments/files/19734118/diff.zip) Looking at your Github actions: looks like you build the debug one with temurin, but the release one with adopt. If that's indeed the case, might be good to align the two and switching to temurin for the release builds as well?

**Comment by IzzySoft** [comment]
@Wavesonics do we have a chance here? Seems you're kept busy "elsewhere". May I (for orientation) kindly ask what your plans are in this concern?

**Comment by Wavesonics** [comment]
yeah definitely spending time on other projects right now, but I do intend to focus on this soon.

I can align those easy enough, but the Moko stuff I'm still not sure about.

**Comment by IzzySoft** [comment]
We can go one step after the other. Maybe align "all else" first (that which you said is "easy enough") and we see how far it gets us. Maybe Moko turns out to be less an issue as we think – maybe not. But we can see that if all else was done?

**Comment by Wavesonics** [comment]
Ya sounds like a plan, should have some time to focus on it soon.

**Comment by IzzySoft** [comment]
Cool! Looking forward to your ping then when you're ready, so I shall have another run 😃

**Comment by IzzySoft** [comment]
Looks a bit like some sorting issue, at least partly:

![Image](https://github.com/user-attachments/assets/f5a3e89d-6a99-4f39-a69a-3ab40636165c)

So basically "situation unchanged". That was from v0.10.2. All diffs here: [diffs.zip](https://github.com/user-attachments/files/20733293/diffs.zip)

**Comment by IzzySoft** [comment]
@Wavesonics haven't heard back from you for a while – but have checked every now and then. Looks like your latest release is RB now, congrats! So the listing at IzzyOnDroid has been established again. Welcome back!

Closing this issue now. Should things break again, maybe it would be time for a fresh issue then – but hopefully, that won't be needed 😉

**Comment by Wavesonics** [comment]
hooray 🎉
